repos/system_upgrade/common/actors/targetuserspacecreator/tests/unit_test_targetuserspacecreator.py

```diff
@@ -48,6 +50,183 @@ def __exit__(self, exception_type, exception_value, traceback):
         pass
 
 
+def traverse_structure(structure, root=Path('/')):
+    for filename, links_to in structure.items():
+        filepath = root / filename
+
+        if isinstance(links_to, dict):
+            for pair in traverse_structure(links_to, filepath):
+                yield pair
+        else:
+            yield (filepath, links_to)
+
+
+def assert_directory_structure_matches(root, initial, expected):
+    for filepath, links_to in traverse_structure(expected, root=root / 'expected'):
+        assert filepath.exists()
+
+        if links_to is None:
+            assert filepath.is_file()
+            continue
+
+        assert filepath.is_symlink()
+        assert os.readlink(filepath) == str(root / 'initial' / links_to.lstrip('/'))
+
+
+@pytest.fixture
+def temp_directory_layout(tmp_path, initial_structure):
+    for filepath, links_to in traverse_structure(initial_structure, root=tmp_path / 'initial'):
+        file_path = tmp_path / filepath
+        file_path.parent.mkdir(parents=True, exist_ok=True)
+
+        if links_to is None:
+            file_path.touch()
+            continue
+
+        file_path.symlink_to(tmp_path / 'initial' / links_to.lstrip('/'))
+
+    (tmp_path / 'expected').mkdir()
+    assert (tmp_path / 'expected').exists()
+
+    return tmp_path
+
+
+# The semantics of initial_structure and expected_structure are as follows:
+#
+# 1. The outermost dictionary encodes the root of a directory structure
+#
+# 2. Depending on the value for a key in a dict, each key in the dictionary
+#    denotes the name of either a:
+#        a) directory -- if value is dict
+#        b) regular file -- if value is None
+#        c) symlink -- if a value is str
+#
+# 3. The value of a symlink entry is a absolute path to a file in the context of
+#    the structure.
+#
```

Comment:
We also need A => B, B => C, C => A (expectation: broken symlink so none of the three files is present)

Reply:
Sure, totally forgot about this possibility. Adding!